File: src/graphics/window.cpp
#include "window.h"
#include <SDL3/SDL_events.h>
#include <SDL3/SDL_stdinc.h>
#include <SDL3/SDL_vulkan.h>
#include <iostream>

void sdl_panic() {
  std::cout << "[SDL] " << SDL_GetError() << std::endl;
  exit(1);
}

SDL_Window *create_window(const std::string title) {
  SDL_Window *window = SDL_CreateWindow(
      title.c_str(), 800, 600, SDL_WINDOW_VULKAN | SDL_WINDOW_RESIZABLE);
  if (window == nullptr) {
    sdl_panic();
  }

  return window;
}

Window::Window(const std::string title) {
  if (!SDL_Init(0)) {
    sdl_panic();
  }

  window = create_window(title);
}

Window::~Window() {
  SDL_DestroyWindow(window);
  SDL_Quit();
}

VkSurfaceKHR Window::surface(const VkInstance vk_instance) const {
  VkSurfaceKHR surface;
  SDL_Vulkan_CreateSurface(window, vk_instance, nullptr, &surface);

  return surface;
}

std::vector<const char *> Window::vulkan_instance_extensions() const {
  Uint32 extension_count;
  char const *const *extensions =
      SDL_Vulkan_GetInstanceExtensions(&extension_count);

  std::vector<const char *> extensions_vector;
  for (Uint32 i = 0; i < extension_count; i++) {
    extensions_vector.push_back(extensions[i]);
  }

  return extensions_vector;
}

void Window::start_event_loop(Graphics &graphics) const {
  bool is_running = true;
  while (is_running) {
    SDL_Event event;
    while (SDL_PollEvent(&event)) {
      switch (event.type) {
      case SDL_EVENT_WINDOW_CLOSE_REQUESTED:
        is_running = false;
        break;
      case SDL_EVENT_WINDOW_RESIZED:
        graphics.should_recreate_swapchain = true;
        break;
      case SDL_EVENT_KEY_DOWN:
        if (event.key.key == SDLK_ESCAPE) {
          is_running = false;
        }
        break;
      }
    }

    if (graphics.should_recreate_swapchain) {
      graphics.recreate_rendering();
    }
    graphics.render();
  }
}

File: src/window.zig
const builtin = @import("builtin");
const sdl = @import("sdl");
const std = @import("std");

const util = @import("util.zig");

const Color = @import("color.zig").Color;

pub const Window = struct {
    sdl_window: *sdl.SDL_Window,

    clear_color: Color,

    pub fn init(window_title: []const u8, clear_color: Color) Window {
        if (!sdl.SDL_SetHint(sdl.SDL_HINT_RENDER_DRIVER, "vulkan")) util.sdlPanic();
        if (!sdl.SDL_Init(sdl.SDL_INIT_VIDEO)) {
            util.sdlPanic();
        }

        const sdl_window: *sdl.SDL_Window = sdl.SDL_CreateWindow(
            window_title.ptr,
            1280,
            720,
            sdl.SDL_WINDOW_RESIZABLE,
        ) orelse util.sdlPanic();

        return Window{
            .sdl_window = sdl_window,
            .clear_color = clear_color,
        };
    }

    pub fn deinit(_: *Window) void {
        sdl.SDL_Quit();
    }

    pub fn readEvent(_: *const Window) ?sdl.SDL_Event {
        var event: sdl.SDL_Event = undefined;
        if (sdl.SDL_PollEvent(&event)) {
            return event;
        }

        return null;
    }
};
